CalibMuon/DTCalibration/plugins/DTResidualHistory.cc

```diff
@@ -0,0 +1,308 @@
+
+/*
+ *  See header file for a description of this class.
+ *
+ */
+
+#include "DTResidualHistory.h"
+
+// Framework
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+//Geometry
+#include "Geometry/DTGeometry/interface/DTGeometry.h"
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+
+//RecHit
+#include "DataFormats/DTRecHit/interface/DTRecSegment4DCollection.h"
+#include "DataFormats/DTRecHit/interface/DTRecHitCollection.h"
+
+#include "CommonTools/Utils/interface/TH1AddDirectorySentry.h"
+#include "CalibMuon/DTCalibration/interface/DTSegmentSelector.h"
+#include "CalibMuon/DTCalibration/interface/DTRecHitSegmentResidual.h"
+
+#include "TFile.h"
+#include "TH1F.h"
+#include "TH2F.h"
+
+#include <algorithm>
+
+using namespace std;
+
+DTResidualHistory::DTResidualHistory(const edm::ParameterSet& pset):
+  select_(pset),
+  segment4DLabel_(pset.getParameter<edm::InputTag>("segment4DLabel")),
+  rootBaseDir_(pset.getUntrackedParameter<std::string>("rootBaseDir","DT/Residuals")),
+  detailedAnalysis_(pset.getUntrackedParameter<bool>("detailedAnalysis",false)) {
+
+  edm::LogVerbatim("Calibration") << "[DTResidualHistory] Constructor called.";
+  consumes< DTRecSegment4DCollection >(edm::InputTag(segment4DLabel_));
+  std::string rootFileName = pset.getUntrackedParameter<std::string>("rootFileName","residuals.root");
+  rootFile_ = new TFile(rootFileName.c_str(), "RECREATE");
+  rootFile_->cd();
+  
+  segmok=0;
+  segmbad=0;
+  nevent=0;
+}
+
+DTResidualHistory::~DTResidualHistory() {
+  edm::LogVerbatim("Calibration") << "[DTResidualHistory] Destructor called.";
+  cout << endl << " Finished. " << endl;
+  cout << " Analyzed events: " << nevent << endl;
+  cout << " Good segments: " << segmok << endl;
+  cout << " Bad segments: " << segmbad << endl;
+}
+
+void DTResidualHistory::beginJob() {
+//  TH1::SetDefaultSumw2(true);
+}
+
+void DTResidualHistory::beginRun(const edm::Run& run, const edm::EventSetup& setup) {
+  
+  // get the geometry
+  edm::ESHandle<DTGeometry> dtGeomH; 
+  setup.get<MuonGeometryRecord>().get(dtGeomH);
+  dtGeom_ = dtGeomH.product();
+  lastrun = 0;
+}
+
+void DTResidualHistory::analyze(const edm::Event& event, const edm::EventSetup& setup) {
+
+  const DTSuperLayerId slIdToFill(-2,1,1,1);
+
+  rootFile_->cd();
+  ++nevent;
+  unsigned int run = event.id().run();
+  if (run!=lastrun) {
+     TH1AddDirectorySentry addDir;
+     histoMapTH1F_.clear();
+     auto ch_it = dtGeom_->chambers().begin(); 	 
+     auto ch_end = dtGeom_->chambers().end(); 	 
+     for (; ch_it != ch_end; ++ch_it) { 	 
```

Comment:
could these three loops be simplified into range based loops?

Reply:
The comments on DTResidualHistory.cc apply also to other code in CalibMuon/DTCalibration/plugins - I guess this file was reviewed now because it was new. The loops above are copied over from DTResidualCalibration.cc, as is the histogram booking. std::stringstream is used in ~100 places across the whole directory. 
If these are to be fixed then I guess it makes sense to fix them everywhere, but its going to take a while

Reply:
indeed - its a big job to update all legacy code - but that is not what is asked - rather we would like new code to be updated. Thanks

Reply:
well I did do a slightly more extensive fix than just the new code, at least concerning loops and std::to_string.